### ISSUE TITLE
compiler warnings with clang: atomics problem?

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#6264]
Per https://www.mail-archive.com/devel@lists.open-mpi.org/msg20843.html, @ggouaillardet discovered that clang 5 is reporting some serious-sounding warnings:

```
  CC       opal_convertor_raw.lo
In file included from opal_convertor_raw.c:21:
In file included from ../../opal/datatype/opal_convertor_internal.h:21:
In file included from ../../opal/datatype/opal_convertor.h:35:
In file included from ../../opal/datatype/opal_datatype.h:43:
In file included from ../../opal/class/opal_object.h:126:
In file included from ../../opal/threads/thread_usage.h:30:
In file included from ../../opal/include/opal/sys/atomic.h:63:
../../opal/include/opal/sys/atomic_stdc.h:109:1: error: address argument to atomic operation must be a pointer to integer or pointer ('opal_atomic_int32_t *' (aka '_Atomic(int32_t) *') invalid)
OPAL_ATOMIC_STDC_DEFINE_FETCH_OP(add, 32, int32_t, +)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../opal/include/opal/sys/atomic_stdc.h:101:16: note: expanded from macro 'OPAL_ATOMIC_STDC_DEFINE_FETCH_OP'
        return atomic_fetch_ ## op ## _explicit (addr, value, memory_order_relaxed); \
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I cannot replicate these errors on Cisco systems with clang 5, 6, or 7.

**Comment by jsquyres** [comment]
Correction: the clang errors I'm seeing on Cisco systems are *different* than the errors @ggouaillardet reported.  I cannot replicate the errors that @ggouaillardet reported -- perhaps I'm not using the same compiler flags...?

**Comment by ggouaillardet** [comment]
clang 5.0 on trusty appears to be busted.

```
$ cat /etc/apt/sources.list.d/llvm.list 
deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-5.0 main
```

this simple program can be used to evidence the issue

```c
#include <stdatomic.h>
#include <stdint.h>

void conftest() {
    _Atomic uint32_t a;
    uint32_t b;
    atomic_fetch_xor_explicit(&a, b, memory_order_relaxed);
}
```

with clang 5.0

```
$ clang-5.0 -c conftest.c 
conftest.c:7:5: error: address argument to bitwise atomic operation must be a
      pointer to integer ('_Atomic(uint32_t) *' invalid)
    atomic_fetch_xor_explicit(&a, b, memory_order_relaxed);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/i686-linux-gnu/4.9/include/stdatomic.h:224:6: note: 
      expanded from macro 'atomic_fetch_xor_explicit'
                          __atomic_fetch_xor ((PTR), (VAL), (MO))
                          ^                   ~~~~~
1 error generated.
```

but the same test works just fine with clang 6.0 and clang 7

#6272 enhances the C11 detection and disables it if this test does not pass.

**Comment by jsquyres** [comment]
#6272 was merged -- do you want to close this issue?

**Comment by ggouaillardet** [comment]
yep, let's close it (and reopen it if is is nor merged in `v4.0.x`)

**Comment by jsquyres** [comment]
Ah, I missed #6723 (merge to v4.0.x).  I'll just add the "Waiting for release branch" label here.

**Comment by awlauria** [comment]
Closing since the reported issues were merged to release branches (#6723).
